modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1920 @@
+            return new GridFinishedFuture<>(new IgniteCheckedException("Snapshot manager is stopping"));
+
+        File nodeSnpDir = null;
+
+        try {
+            String dbNodePath = relativeStoragePath();
+            nodeSnpDir = U.resolveWorkDirectory(new File(tmpWorkDir, snpName).getAbsolutePath(), dbNodePath, false);
+
+            sctx = new LocalSnapshotContext(snpName,
+                nodeSnpDir,
+                parts,
+                exec,
+                snpSndr);
+
+            final LocalSnapshotContext sctx0 = sctx;
+
+            sctx.snpFut.listen(f -> {
+                localSnpCtxs.remove(snpName);
+
+                closeSnapshotResources(sctx0);
+            });
+
+            for (Map.Entry<Integer, GridIntList> e : parts.entrySet()) {
+                final CacheGroupContext gctx = cctx.cache().cacheGroup(e.getKey());
+
+                // Create cache snapshot directory if not.
+                File grpDir = U.resolveWorkDirectory(sctx.nodeSnpDir.getAbsolutePath(),
+                    cacheDirName(gctx.config()), false);
+
+                U.ensureDirectory(grpDir,
+                    "bakcup directory for cache group: " + gctx.groupId(),
+                    null);
+
+                CompletableFuture<Boolean> cpEndFut0 = sctx.cpEndFut;
+
+                GridIntIterator iter = e.getValue().iterator();
+
+                while (iter.hasNext()) {
+                    int partId = iter.next();
+
+                    GroupPartitionId pair = new GroupPartitionId(e.getKey(), partId);
+                    PageStore store = ((FilePageStoreManager)cctx.pageStore()).getStore(pair.getGroupId(),
+                        pair.getPartitionId());
+
+                    sctx.partDeltaWriters.put(pair,
+                        new PageStoreSerialWriter(log,
+                            store,
+                            () -> cpEndFut0.isDone() && !cpEndFut0.isCompletedExceptionally(),
+                            sctx.snpFut,
+                            getPartionDeltaFile(grpDir, partId),
+                            ioFactory,
+                            pageSize));
+                }
+            }
+
+            LocalSnapshotContext ctx0 = localSnpCtxs.putIfAbsent(snpName, sctx);
+
+            assert ctx0 == null : ctx0;
+
+            CheckpointFuture cpFut = dbMgr.forceCheckpoint(String.format(SNAPSHOT_CP_REASON, snpName));
+
+            cpFut.finishFuture()
+                .listen(f -> {
+                    if (f.error() == null)
+                        sctx0.cpEndFut.complete(true);
+                    else
+                        sctx0.cpEndFut.completeExceptionally(f.error());
+                });
+
+            cpFut.beginFuture()
+                .get();
+
+            if (log.isInfoEnabled())
+                log.info("Snapshot operation scheduled with the following context: " + sctx);
+        }
+        catch (IOException e) {
+            closeSnapshotResources(sctx);
```

Comment:
Complete `snpFut` here, instead of closing snapshot resources.

Reply:
Removed as not actual.